Homework 4

JeRusakow opened the pull request.

JeRusakow committed:
Task1 solved

JeRusakow committed:
Task2 solved

JeRusakow committed:
Error name fix

JeRusakow committed:
Task 3 solved

JeRusakow committed:
Doc

JeRusakow committed:
Docstrings added

JeRusakow committed:
Task4 solved. Doctest instruction provided

JeRusakow committed:
FizzBuzz now supports numbers divisivle by 15

JeRusakow committed:
Task5 solved. Task4 modified

Also added test for a number divisible by both 3 and 5 task4

iamthen0ise commented on homework4/task1/task_1_read_file.py:
```diff
@@ -0,0 +1,51 @@
+
+    datafile = open(path)
+    line = datafile.readline()
+    datafile.close()
```
но ведь можно `with open...` и не придется закрывать файл самому.

JeRusakow replied:
Переписал с использованием менеджера контекста

iamthen0ise commented on homework4/task1/task_1_read_file.py:
```diff
@@ -0,0 +1,51 @@
+    print(f"Line: '{line}'")
+
+    try:
+        num = float(line)
```
кажется без целей дебага эта переменная используется 1 раз. конструкция Try позволяет не делать одноразовые переменные, а прям нагло проверять все на месте)

JeRusakow replied:
Удалил остатки отладочного вывода. Теперь код выглядит лаконичней :)

iamthen0ise commented on homework4/task2/task_2_mock_input.py:
```diff
@@ -0,0 +1,45 @@
+        raise ValueError(f"Unreachable {url}")
+
+    if response.status != 200:
+        raise ValueError(f"Unreachable {url}")
```
но вель Unreachable URL это слишком абстрактно для разных ошибок. Думаю, ты можешь разнообразить экзепшены. Пусть будет показана причина падения, важно понимать — когда нет интернета, а когда 400/500 ошибки.

JeRusakow replied:
Изменил сообщение при втором исключении на "Bad response code: XXX". Пусть их будет два типа

(Пока из интереса смотрел список возможных кодов ответа, наткнулся на забавный номер 418)))

iamthen0ise commented on homework4/task4/README.md:
```diff
@@ -0,0 +1,11 @@
+<h2>Running doctests for the module with pytest</h2>
```
Ридми — всегда круто! 👍

JeRusakow replied:
Прочитал мануал по местному диалекту разметки и переписал)

iamthen0ise commented on homework4/task5/task_5_optional.py:
```diff
@@ -0,0 +1,34 @@
+
+def fizzbuzz(n: int) -> Generator[str, None, None]:
+    """Generates a Fizz-Buzz sequence of length n"""
+    fizzes = cycle(["", "", "Fizz"])
```
мощно, а без cycle получится?) это на интерес вопрос) если нет — так нет

JeRusakow replied:
Получилось)
Ажно сам удивился, насколько просто открывался ларчик

iamthen0ise commented on tests/test_homework4/test_get_print_output.py:
```diff
@@ -0,0 +1,17 @@
+
+
+@patch("sys.stderr.write")
+def test_stderr_write(fake_stderr):
```
что-то тут не так. в тестах я поменял весь инпут на `abc` и оба теста прошли.
думаю стоит добавить тест с абсолютно рандомным текстом, можно даже автогенерируемым типа  `"".join([chr(random.randint(50,120)) for i in range(10)])`

JeRusakow replied:
Добавил тест с автогенерируемым сообщением.
Оказалось, тесты проходили, потому что assert тестировал совсем не то: проверялось само создание атрибутов у заглушки вместо истории вызовов. 
Было `assert mock.called_with()` вместо `mock.assert_called_with()`

iamthen0ise reviewed (changes requested):
что-то не то с задачей на вывод в потоки, в остальном только поправить по комментариям

JeRusakow committed:
open/close --> with open as @ task1

Also deletet debug prints

JeRusakow committed:
Changed exception description for non-200 HTTP Responses.

Now it says "Bad response code: XXX"

JeRusakow committed:
Random message test is introdused. Tests suite strictened.

JeRusakow committed:
Found the reason of test false-passing

Asserted was the creation of new mock attributes (>>> assert mock.called_once())
instead of real call assertion (>>> mock.assert_called_once())

JeRusakow committed:
Fizz-Buzz generator rewritten not to use itertools.cycle.

More than that, the solution now looks cleaner and simplier because of using fewer entities.

JeRusakow committed:
README now better corresponds to Github markdown language

I should have really read github markdown manual
